File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        int l = 0, r = nums.size() - 1;
        while (r - l > 1) {
            int m = (r + l) / 2;
            if(nums[l] <= nums[m] && nums[m] <= nums[r]) return nums[l];
            else if(nums[l] < nums[m]) l = m + 1;
            else if(nums[m] < nums[r]) r = m;
        }
        return min(nums[r], nums[l]);
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        bestSoFar = nums[0]
        maxProd = 1
        minProd = 1
        for num in nums:
            temp = max(minProd * num, maxProd * num, num)
            minProd = min(minProd * num, maxProd * num, num)
            maxProd = temp
            bestSoFar = max(bestSoFar, maxProd)
            print(num, minProd, maxProd, bestSoFar)
        return bestSoFar

        